PairwiseAligner report output is local dependent

Hello,

More a question than an issue: while running the tests with local set to French (side effect of working on #4972), I saw some tests do not pass because french use dot instead of comma as decimal separator (an subset of the error hereafter). I don't know if that could be an issue as the output seems very user oriented to me, while a programme could also easily parse it.


```log
(...)
======================================================================
FAIL: test_gap_here_only_local_2 (test_pairwise_aligner.TestPerSiteGapPenalties.test_gap_here_only_local_2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/brancotte/projects/biopython/Tests/test_pairwise_aligner.py", line 4614, in test_gap_here_only_local_2
    self.assertEqual(
AssertionError: 'Pair[68 chars]re: 1,000000\n  mismatch_score: -1,000000\n  i[263 chars]al\n' != 'Pair[68 chars]re: 1.000000\n  mismatch_score: -1.000000\n  i[263 chars]al\n'
  Pairwise sequence aligner with parameters
    wildcard: None
-   match_score: 1,000000
?                 ^
+   match_score: 1.000000
?                 ^
-   mismatch_score: -1,000000
?                     ^
+   mismatch_score: -1.000000
?                     ^
    insertion_score_function: <function TestPerSiteGapPenalties.test_gap_here_only_local_2.<locals>.nogaps at 0x742edb51e660>
    deletion_score_function: <function TestPerSiteGapPenalties.test_gap_here_only_local_2.<locals>.specificgaps at 0x742edb51c220>
    mode: local


======================================================================
FAIL: test_gap_here_only_local_3 (test_pairwise_aligner.TestPerSiteGapPenalties.test_gap_here_only_local_3)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/brancotte/projects/biopython/Tests/test_pairwise_aligner.py", line 4723, in test_gap_here_only_local_3
    self.assertEqual(
AssertionError: 'Pair[68 chars]re: 1,000000\n  mismatch_score: -10,000000\n  [381 chars]al\n' != 'Pair[68 chars]re: 1.000000\n  mismatch_score: -10.000000\n  [381 chars]al\n'
Diff is 1341 characters long. Set self.maxDiff to None to see it.

--
```

Hi @mdehoon 

Simply add `locale.setlocale(locale.LC_ALL, "fr_FR.utf8")` at any moment in test such as test_gap_here_only_local_2 and it will crash. Note that you have to have french locale one your computer. Hereafter a consuming-bandwidth minimal working example

```sh
docker run -it python:3.12 bash
apt-get update && apt-get install -y locales
echo "fr_FR.UTF-8 UTF-8" >> /etc/locale.gen
locale-gen
locale -a # check that all three lang are not there
git clone https://github.com/biopython/biopython.git
cd biopython
pip install -e .
pip install -r ci-dependencies.txt
cd Tests
# ok
python run_tests.py test_pairwise_aligner.TestPerSiteGapPenalties.test_gap_here_only_local_2
# force to use french ...
sed -i '1 i\import locale' test_pairwise_aligner.py
sed -i 's/test_gap_here_only_local_2(self):/&\n        locale.setlocale(locale.LC_ALL, "fr_FR.utf8")/' test_pairwise_aligner.py
# ... and redo the test
python run_tests.py test_pairwise_aligner.TestPerSiteGapPenalties.test_gap_here_only_local_2
```

resulting in

```log
root@4d8a500a026f:/biopython/Tests# python run_tests.py test_pairwise_aligner.TestPerSiteGapPenalties.test_gap_here_only_local_2
Python version: 3.12.7 (main, Oct 19 2024, 04:14:06) [GCC 12.2.0]
Operating system: posix linux
/biopython/Bio/__init__.py:138: BiopythonWarning: You may be importing Biopython from inside the source tree. This is bad practice and might lead to downstream issues. In particular, you might encounter ImportErrors due to missing compiled C extensions. We recommend that you try running your code from outside the source tree. If you are outside the source tree then you have a setup.py file in an unexpected directory: /biopython
  warnings.warn(
test_pairwise_aligner.TestPerSiteGapPenalties.test_gap_here_only_local_2 ... ok
----------------------------------------------------------------------
Ran 1 test in 0.202 seconds

root@4d8a500a026f:/biopython/Tests# 
root@4d8a500a026f:/biopython/Tests# sed -i '1 i\import locale' test_pairwise_aligner.py
root@4d8a500a026f:/biopython/Tests# sed -i 's/test_gap_here_only_local_2(self):/&\n        locale.setlocale(locale.LC_ALL, "fr_FR.utf8")/' test_pairwise_aligner.py
root@4d8a500a026f:/biopython/Tests# python run_tests.py test_pairwise_aligner.TestPerSiteGapPenalties.test_gap_here_only_local_2
Python version: 3.12.7 (main, Oct 19 2024, 04:14:06) [GCC 12.2.0]
Operating system: posix linux
/biopython/Bio/__init__.py:138: BiopythonWarning: You may be importing Biopython from inside the source tree. This is bad practice and might lead to downstream issues. In particular, you might encounter ImportErrors due to missing compiled C extensions. We recommend that you try running your code from outside the source tree. If you are outside the source tree then you have a setup.py file in an unexpected directory: /biopython
  warnings.warn(
test_pairwise_aligner.TestPerSiteGapPenalties.test_gap_here_only_local_2 ... FAIL
======================================================================
FAIL: test_gap_here_only_local_2 (test_pairwise_aligner.TestPerSiteGapPenalties.test_gap_here_only_local_2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/biopython/Tests/test_pairwise_aligner.py", line 4616, in test_gap_here_only_local_2
    self.assertEqual(
AssertionError: 'Pair[68 chars]re: 1,000000\n  mismatch_score: -1,000000\n  i[263 chars]al\n' != 'Pair[68 chars]re: 1.000000\n  mismatch_score: -1.000000\n  i[263 chars]al\n'
  Pairwise sequence aligner with parameters
    wildcard: None
-   match_score: 1,000000
?                 ^
+   match_score: 1.000000
?                 ^
-   mismatch_score: -1,000000
?                     ^
+   mismatch_score: -1.000000
?                     ^
    insertion_score_function: <function TestPerSiteGapPenalties.test_gap_here_only_local_2.<locals>.nogaps at 0x79a480471260>
    deletion_score_function: <function TestPerSiteGapPenalties.test_gap_here_only_local_2.<locals>.specificgaps at 0x79a480470400>
    mode: local


----------------------------------------------------------------------
Ran 1 test in 0.080 seconds

FAILED (failures = 1)
root@4d8a500a026f:/biopython/Tests# pip freeze | grep biopython
-e git+https://github.com/biopython/biopython.git@180d8a6e69fbbdc7e3db52122a18a7e3a306b1e6#egg=biopython
root@4d8a500a026f:/biopython/Tests# 

```

--- a/Bio/Align/_pairwisealigner.c
+++ b/Bio/Align/_pairwisealigner.c
@@ -1961,11 +1961,12 @@ Aligner_str(Aligner* self)
 {
     char text[1024];
     char* p = text;
+    char* value;
     PyObject* substitution_matrix = self->substitution_matrix.obj;
     void* args[3];
     int n = 0;
     PyObject* wildcard = NULL;
-    PyObject* s;
+    PyObject* s = NULL;
 
     p += sprintf(p, "Pairwise sequence aligner with parameters\n");
     if (substitution_matrix) {
@@ -1983,44 +1984,88 @@ Aligner_str(Aligner* self)
             p += sprintf(p, "  wildcard: '%%U'\n");
             args[n++] = wildcard;
         }
-        p += sprintf(p, "  match_score: %f\n", self->match);
-        p += sprintf(p, "  mismatch_score: %f\n", self->mismatch);
+        /* Use PyOS_double_to_string to ensure that the locale does
+         * not change the decimal point into a comma.
+         */
+        value = PyOS_double_to_string(self->match, 'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  match_score: %s\n", value);
+        PyMem_Free(value);
+        value = PyOS_double_to_string(self->mismatch, 'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  mismatch_score: %s\n", value);
+        PyMem_Free(value);
     }
     if (self->insertion_score_function) {
         p += sprintf(p, "  insertion_score_function: %%R\n");
         args[n++] = self->insertion_score_function;
     }
     else {
-        p += sprintf(p, "  open_internal_insertion_score: %f\n",
-                     self->open_internal_insertion_score);
-        p += sprintf(p, "  extend_internal_insertion_score: %f\n",
-                     self->extend_internal_insertion_score);
-        p += sprintf(p, "  open_left_insertion_score: %f\n",
-                     self->open_left_insertion_score);
-        p += sprintf(p, "  extend_left_insertion_score: %f\n",
-                     self->extend_left_insertion_score);
-        p += sprintf(p, "  open_right_insertion_score: %f\n",
-                     self->open_right_insertion_score);
-        p += sprintf(p, "  extend_right_insertion_score: %f\n",
-                     self->extend_right_insertion_score);
+        value = PyOS_double_to_string(self->open_internal_insertion_score,
+                                      'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  open_internal_insertion_score: %s\n", value);
+        PyMem_Free(value);
+        value = PyOS_double_to_string(self->extend_internal_insertion_score,
+                                      'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  extend_internal_insertion_score: %s\n", value);
+        PyMem_Free(value);
+        value = PyOS_double_to_string(self->open_left_insertion_score,
+                                      'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  open_left_insertion_score: %s\n", value);
+        PyMem_Free(value);
+        value = PyOS_double_to_string(self->extend_left_insertion_score,
+                                      'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  extend_left_insertion_score: %s\n", value);
+        PyMem_Free(value);
+        value = PyOS_double_to_string(self->open_right_insertion_score,
+                                      'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  open_right_insertion_score: %s\n", value);
+        PyMem_Free(value);
+        value = PyOS_double_to_string(self->extend_right_insertion_score,
+                                      'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  extend_right_insertion_score: %s\n", value);
+        PyMem_Free(value);
     }
     if (self->deletion_score_function) {
         p += sprintf(p, "  deletion_score_function: %%R\n");
         args[n++] = self->deletion_score_function;
     }
     else {
-        p += sprintf(p, "  open_internal_deletion_score: %f\n",
-                     self->open_internal_deletion_score);
-        p += sprintf(p, "  extend_internal_deletion_score: %f\n",
-                     self->extend_internal_deletion_score);
-        p += sprintf(p, "  open_left_deletion_score: %f\n",
-                     self->open_left_deletion_score);
-        p += sprintf(p, "  extend_left_deletion_score: %f\n",
-                     self->extend_left_deletion_score);
-        p += sprintf(p, "  open_right_deletion_score: %f\n",
-                     self->open_right_deletion_score);
-        p += sprintf(p, "  extend_right_deletion_score: %f\n",
-                     self->extend_right_deletion_score);
+        value = PyOS_double_to_string(self->open_internal_deletion_score,
+                                      'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  open_internal_deletion_score: %s\n", value);
+        PyMem_Free(value);
+        value = PyOS_double_to_string(self->extend_internal_deletion_score,
+                                      'f', 6, 0, NULL);
+        p += sprintf(p, "  extend_internal_deletion_score: %s\n", value);
+        PyMem_Free(value);
+        value = PyOS_double_to_string(self->open_left_deletion_score,
+                                      'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  open_left_deletion_score: %s\n", value);
+        PyMem_Free(value);
+        value = PyOS_double_to_string(self->extend_left_deletion_score,
+                                      'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  extend_left_deletion_score: %s\n", value);
+        PyMem_Free(value);
+        value = PyOS_double_to_string(self->open_right_deletion_score,
+                                      'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  open_right_deletion_score: %s\n", value);
+        PyMem_Free(value);
+        value = PyOS_double_to_string(self->extend_right_deletion_score,
+                                      'f', 6, 0, NULL);
+        if (!value) goto exit;
+        p += sprintf(p, "  extend_right_deletion_score: %s\n", value);
+        PyMem_Free(value);
     }
     switch (self->mode) {
         case Global: sprintf(p, "  mode: global\n"); break;
@@ -2031,6 +2076,8 @@ Aligner_str(Aligner* self)
             return NULL;
     }
     s = PyUnicode_FromFormat(text, args[0], args[1], args[2]);
+
+exit:
     Py_XDECREF(wildcard);
     return s;
 }

--- a/Bio/Cluster/clustermodule.c
+++ b/Bio/Cluster/clustermodule.c
@@ -942,12 +942,13 @@ PyTree_str(PyTree* self)
     PyObject* temp;
 
     output = PyUnicode_FromString("");
+    if (!output) return NULL;
+
     for (i = 0; i < n; i++) {
         node = self->nodes[i];
         distance = PyOS_double_to_string(node.distance, 'g', 6, 0, NULL);
         if (!distance) {
             Py_DECREF(output);
-            Py_DECREF(line);
             return NULL;
         }
         /* Use PyOS_double_to_string to ensure that the locale does
@@ -962,11 +963,9 @@ PyTree_str(PyTree* self)
             return NULL;
         }
         temp = PyUnicode_Concat(output, line);
-        if (!temp) {
-            Py_DECREF(output);
-            Py_DECREF(line);
-            return NULL;
-        }
+        Py_DECREF(line);
+        Py_DECREF(output);
+        if (!temp) return NULL;
         output = temp;
     }
     return output;

--- a/Bio/Cluster/clustermodule.c
+++ b/Bio/Cluster/clustermodule.c
@@ -682,8 +682,15 @@ PyNode_repr(PyNode* self)
 {
     char string[64];
 
-    sprintf(string, "(%d, %d): %g",
-                   self->node.left, self->node.right, self->node.distance);
+    /* Use PyOS_double_to_string to ensure that the locale does
+     * not change the decimal point into a comma.
+     */
+    char* value = PyOS_double_to_string(self->node.distance, 'g', 6, 0, NULL);
+    if (!value) return NULL;
+
+    sprintf(string, "(%d, %d): %s", self->node.left, self->node.right, value);
+    PyMem_Free(value);
+
     return PyUnicode_FromString(string);
 }
 
@@ -928,6 +935,7 @@ PyTree_str(PyTree* self)
     int i;
     const int n = self->n;
     char string[128];
+    char* distance;
     Node node;
     PyObject* line;
     PyObject* output;
@@ -936,7 +944,17 @@ PyTree_str(PyTree* self)
     output = PyUnicode_FromString("");
     for (i = 0; i < n; i++) {
         node = self->nodes[i];
-        sprintf(string, "(%d, %d): %g", node.left, node.right, node.distance);
+        distance = PyOS_double_to_string(node.distance, 'g', 6, 0, NULL);
+        if (!distance) {
+            Py_DECREF(output);
+            Py_DECREF(line);
+            return NULL;
+        }
+        /* Use PyOS_double_to_string to ensure that the locale does
+         * not change the decimal point into a comma.
+         */
+        sprintf(string, "(%d, %d): %s", node.left, node.right, distance);
+        PyMem_Free(distance);
         if (i < n-1) strcat(string, "\n");
         line = PyUnicode_FromString(string);
         if (!line) {

